### ISSUE TITLE
Size error at run-time

### ISSUE THREAD

**Issue by athas** [diku-dk/futhark#1042]
```futhark
-- ==
-- input { [0,1,2,3] }

let writeSegment 'a [m] [n] (dest: *[m]a) startIndex (vs: [n]a) : *[m]a =
  let is = map (+startIndex) (iota n)
  in scatter dest is vs

let main [n] (indices: [n]i32) =
  let cs = replicate (n*(n-1)/2) (0,0)
  let (cs, _) = loop (cs, j)=(cs, 0)
                for i < (n-1) do
                let segment = map (\k -> (indices[j],k)) <| drop (j+1) indices
                in (writeSegment cs j segment, j+n-i-1)
  in cs

```

Looks like a type checking bug.

**Comment by athas** [comment]
I think this program should not actually be type correct.  It violates causality.

**Comment by athas** [comment]
Closed in 71130df2f01be68bb411a90c98968d9f2c203fc7.
